File: code/23_solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def mergeKLists(self, lists: List[ListNode]) -> ListNode:
        if not lists: return 
        
        def merge(l1, l2):
            ret = curr = ListNode(None)
            while l1 and l2:
                if l1.val < l2.val:
                    curr.next = l1
                    l1 = l1.next
                else:
                    curr.next = l2
                    l2 = l2.next
                curr = curr.next
            curr.next = l1 if l1 else l2
            return ret.next
        
        lists = collections.deque(lists)
        while len(lists) > 1:
            l1 = lists.popleft()
            l2 = lists.popleft()
            l3 = merge(l1, l2)
            lists.append(l3)
        return lists.pop()

File: code/23_solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
private:
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        ListNode ret(0);
        ListNode* trav = &ret;
        while (l1 and l2){
            if (l1->val < l2->val) {
                trav->next = l1;
                l1 = l1->next;
            } else {
                trav->next = l2;
                l2 = l2->next;                
            }
            trav = trav->next;
        }
        trav->next = l1 ? l1 : l2;
        return ret.next;
    }
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        queue<ListNode*> q;
        for (auto node:lists) q.push(node);
        while (q.size() > 1) {
            ListNode* l1 = q.front(); q.pop();
            ListNode* l2 = q.front(); q.pop();
            ListNode* l3 = mergeTwoLists(l1, l2);
            q.push(l3);
        }
        return q.empty() ? nullptr: q.front();
    }
};
